### ISSUE TITLE
Turning the Gerber Fail

### ISSUE THREAD

**Issue by mikhailmihalkov** [gerbv/gerbv#134]
If you rotate the file 90 degrees, you cannot save the result. When you reopen, the file returns to its original position. When you change the coordinates or rotate the file, the settings window cannot be moved and it closes the original picture

**Comment by mikhailmihalkov** [comment]
I have tested on almost all versions for win since version 2.6.and does not work on linux 2.6.1 either.

**Comment by mikhailmihalkov** [comment]
everything older than these versions unfortunately does not work rotation

**Comment by eyal0** [comment]
I just tested it and it works fine for me.  Maybe you could be more explicit about what you tried and what your expectation?  I will show you what I did:


1. Download a gerber file: `wget https://raw.githubusercontent.com/pcb2gcode/pcb2gcode/master/testing/gerbv_example/example_board_al_custom/example_board-F.Cu.gbr`
2. Type `gerbv example_board-F.Cu.gbr` to view that it says the text "Front".
3. Close gerbv
4. Now load with rotation by typing `gerbv -r 45 example_board-F.Cu.gbr`
5. Observe that the text is now at an angle.
6. In the GUI, select File->Save active layer as...
7. Save the file as `new.gbr`.
8. Close gerbv
9. Now view the new file without any rotation: `gerbv new.gbr`.
10. Observe that rotation was done.

I see that rotation of 45 is working kind of but rotation of a multiple of 90 is being being twice.

Are you trying to rotate by 180?  I see that 180 and 90 are failing for me.

**Comment by mikhailmihalkov** [comment]
Usually I turn 90 degrees and after saving and reopening, I see that the file is back to its original state when viewing

**Comment by mikhailmihalkov** [comment]
Win 10  gerbv 2.9.3~ce0a78
1 Does not save the measurement settings when restarting again switches to mil instead of mm
2 Does not save view settings when restarting switches from normal to fast
3 When rotated 90 degrees and saved, does not flip the image

**Comment by mikhailmihalkov** [comment]
90 Degree Rotation
![image](https://user-images.githubusercontent.com/99812383/198033037-901cedd6-2347-4574-8210-026a37b41111.png)

**Comment by mikhailmihalkov** [comment]
![image](https://user-images.githubusercontent.com/99812383/198034999-f0d215ed-861b-4ceb-ad82-f3fa82104a86.png)

90 Degree Rotation Version 2.6.1

**Comment by mikhailmihalkov** [comment]
I open the saved file in version 2.9.3 with a 90-degree flip in another viewer and see that the file is rotated 180 degrees instead of 90
![1](https://user-images.githubusercontent.com/99812383/198038601-9b9dd5ee-4289-4438-be3f-b3066e1a9e1f.jpg)

**Comment by mikhailmihalkov** [comment]
It turns out that version 2.9.3 visually rotates 90 degrees, but when opened in another program, we see that the image is rotated 180 degrees

**Comment by mikhailmihalkov** [comment]
In version 2.6.1, it visually rotates 180 when choosing 90 degrees, but if you open the file after saving in another program, we will see that the file is rotated 90 (correct)

**Comment by eyal0** [comment]
Such a strange behavior!  We would expect that if you ask for 90 degree rotation, you get a view of 90 degree rotation.  And when you save, it should save it with 90 degree rotation, too!

It's strange to me that the rotation is doubled when the value is not a multiple of 90.  Otherwise, it is not doubled.  I'll investigate a little more.  Maybe this will be a hint that will help us.

**Comment by eyal0** [comment]
I'll only investigate with 2.9.3.  There's no point in trying to fix the old version.

**Comment by eyal0** [comment]
Okay, I think that I see what is happening.

There are two ways to rotate a gerber file.
1. Take all the points in the gerber file and map them to different points.  It's the usually transform using sine and cosine that you learned in high school.  It is described here: https://danceswithcode.net/engineeringnotes/rotations_in_2d/rotations_in_2d.html .  You can run this on each point, no problem.
2. Use the `%IR%` directive.  It is described [here](https://www.ucamco.com/files/downloads/file_en/456/gerber-layer-format-specification-revision-2022-02_en.pdf?7b3ca7f0753aa2d77f5f9afe31b9f826) in section 8.1.5 on page 175.  The only legal rotations are 0, 90, 180, and 270.

gerbv, when you open a file with rotation, is doing *both*!  The points are remapped *and* the `%IR%` directive is added.

After you save the file and re-open it in gerbv then the file is loaded and both are applied.  However, the `%IR%` directive is only valid with rotations of 0, 90, 180, and 270.  So if you choose an invalid rotation, the rotation is ignored and it just opens the files with the points as specified.  And they are rotated already so you'll get an error about the invalid rotation but everything will look rotated as you wanted.

Here's a 30 degree rotation:

![image](https://user-images.githubusercontent.com/109809/198110997-c85f1c86-98e6-45c6-ab19-2577f2fdd106.png)

If you ask for a rotation of 90, you'll have the rotation of the points and the `%IR%` directive but this time, the directive will work!  So you'll end up with double rotation.  That is why the image in https://github.com/gerbv/gerbv/issues/134#issuecomment-1292003457 is rotated 180 degrees.

I tested these files on an online viewer and the invalid rotations didn't even open.  However, the `%IR%` directive is nice because it really minimizes the diff of the file in that it just adds a single line.  I think that it would be nice if gerbv could use the %IR% command when possible and then fall back to doing the rotation of points only if the IR command is not possible.  Then again, it seems that the command has been deprecated since 2021 so maybe we shouldn't be using it at all!

The center of rotation for both the `IR` directive and gerbv's rotation seemed to be 0,0 either way.

Given that the IR command is deprecated, as are all the other, similar, transformation commands, I think that gerbv should never generate one of these.  However, gerbv should continue to accept them.

So the fix is to make it so that, no matter the `-r` input parameter, gerbv will never generate that `%IRxx*%` line.  All the rest of gerbv stays the same.

So:

* If a file is opened **without** `-r` and has an `IR` in it, the IR will stay even if the file is modified and resaved.  The `IR` value is never modified.
* If a file is opened **with** `-r` then the points of the file are rotated about the axis the correct amount.  If the IR statement was previously in there, it stays there.  If it previously wasn't there then it isn't added.  This is true no matter what the IR value was, valid or not.

I think that all of this should be easy to do.  The only thing that we need to change is to make sure that the `-r` command line option does *not* cause any addition or change to the `IR` value.

How does that sound?

**Comment by eyal0** [comment]
`rotation` is kind of a popular word in the codebase so I did a `grep` for `mirrorAroundX`, which is less common.  Looking at the results, it seems like the one that we'd need to fix is here:

https://github.com/gerbv/gerbv/blob/8aa1e02605b01cf8650f58117cf3d0eb81d3b7b7/src/export-rs274x.c#L248-L252

We can see the conversion from degrees to radians and back, that's why 90+60 came out 159 in my file, oops!  Rounding error.  So I'll just adjust this so that instead of addition, we just assign the old value.  It'll still be converting from degrees to radians and back so it might still lose precision.  I'll do a little test but I bet that the problem is that it is truncating to `int` instead of rounding to `int`.  I'll confirm that by adding a little printf and then, if i'm right, I'll change it to rounding in addition to my fix of removing the addition.

All this need unit testing, of course.  I'll add those in the usual way.  This shouldn't take too long.

**Comment by eyal0** [comment]
Yup, I was right about the rounding.  Here's code that I added into that if statement:

```c
                printf("a: %f, b: %f, c: %f, d: %d\n",
                       image->info->imageRotation,
                       thisTransform->rotation,
                       RAD2DEG(image->info->imageRotation + thisTransform->rotation),
                       (int)(RAD2DEG(image->info->imageRotation + thisTransform->rotation)));
```

Here's what was printed:

```
a: 1.570796, b: 1.047198, c: 150.000000, d: 149
```

The file that I opened included this line in it:

```gerber
%IR90*%
```

And I opened it like this:

```bash
gerbv -r 60 new_30_c.gbr
```

So we can see that the first number is the 90 from in the file, converted to radians.  And the second number is the rotation, which is 60 converted to radians.  I'm going to remove that second addition and also add in a rounding procedure.  All the rest of the code stays the same.

Look at that trunaction!  150.0000etc became 149!  So I was right about the rounding.  Rounding to the nearest int should work fine because to get an accurate conversion from degrees to radians and back, you don't need very much precision at all!  The float can do a lot of digits beyond the decimal point and pi/180 is 0.01...  So it ought to be enough.

**Comment by mikhailmihalkov** [comment]
eyal0 Unfortunately, I am not a programmer but an ordinary user) If you do not find it difficult to collect this code into a ready-made version for testing.. For my part, I am ready to test) win and linux

**Comment by eyal0** [comment]
Are you able to compile the code from the sources?  How did you run the code before?

This is the fixed code if you want to test it: https://github.com/eyal0/gerbv/tree/fix_rotation

**Comment by eyal0** [comment]
I added a test to check that it works.  We can see that the original file has an arrow that points up and to the left:

https://github.com/gerbv/gerbv/pull/135/commits/f3a2802ce92ef57df2f0a7274fd8aee09de21905#diff-3fdd1a4b5ad5e84fbd43dc5a708df752e7e2e219930284e6e44a3cc110eeb55c

In the [next revision](https://github.com/gerbv/gerbv/pull/135/commits/709fb910e58b50f0dc7c98b412ec4ea1f324e165), I [added a 90 degree rotation to the test](https://github.com/gerbv/gerbv/pull/135/commits/709fb910e58b50f0dc7c98b412ec4ea1f324e165#diff-b457448a0a285c6bd4c00415b757388ff97268f401e944682a887a54de60cfd8L154-R154).  But in the image, we see that [the arrow was rotated by 180 dergrees](https://github.com/gerbv/gerbv/pull/135/commits/709fb910e58b50f0dc7c98b412ec4ea1f324e165#diff-3fdd1a4b5ad5e84fbd43dc5a708df752e7e2e219930284e6e44a3cc110eeb55c).  It's too much!

In the [revision after that one](https://github.com/gerbv/gerbv/pull/135/commits/82e10644d177b2b53771ca1078730893e33f6cdf), I add [the fix](https://github.com/gerbv/gerbv/pull/135/commits/82e10644d177b2b53771ca1078730893e33f6cdf#diff-4180f9376768ebc7613931d94284e9efe44afb3f6e31c04dbc697c9e38c85c1eL251-R251) and we see that the [arrow is now correctly rotated by only 90 degrees](https://github.com/gerbv/gerbv/pull/135/commits/82e10644d177b2b53771ca1078730893e33f6cdf#diff-3fdd1a4b5ad5e84fbd43dc5a708df752e7e2e219930284e6e44a3cc110eeb55c).

So it seems like it is working.

**Comment by mikhailmihalkov** [comment]
> Умеете ли вы компилировать код из исходников? Как вы запускали код раньше?
> 
> Это фиксированный код, если вы хотите его протестировать: [https://github.com/eyal0/gerbv/tree/fix_rotation .](https://github.com/eyal0/gerbv/tree/fix_rotation)

Unfortunately, I don't know how to compile. On wine I run exe on linux deb package

**Comment by mikhailmihalkov** [comment]
Maybe here? https://gerbv.github.io/#download

**Comment by mikhailmihalkov** [comment]
I use gerbv to cut stencils with this script https://github.com/Jeinzi/SMTCut and I often have to turn the file gerber and remove unnecessary parts

**Comment by eyal0** [comment]
I will try to produce the program for you.  Please wait 1-2 hours for me.  Thanks.

**Comment by eyal0** [comment]
@mikhailmihalkov Can you try this file:

https://github.com/gerbv/gerbv/suites/9000795945/artifacts/414628935

You will need to unzip it and, inside of there, you should find the windows program.

**Comment by mikhailmihalkov** [comment]
> Я постараюсь подготовить программу для вас. Пожалуйста, подождите меня 1-2 часа. Спасибо.



> @mikhailmihalkovМожете ли вы попробовать этот файл:
> 
> https://github.com/gerbv/gerbv/suites/9000795945/artifacts/414628935
> 
> Вам нужно будет разархивировать его, и внутри вы должны найти программу Windows.

many thanks

**Comment by ilya301444** [comment]
@mikhailmihalkov Can you try this file:

https://github.com/gerbv/gerbv/suites/9000795945/artifacts/414628935

You will need to unzip it and, inside of there, you should find the windows program.

you could not build a program for Ubuntu. When I try to run ./gerbv, it asks for the libgerbv.so libraries. I moved the files to the /usr/lib/x86_64-linux-gnu directory and got the error error while loading shared libraries: libdxflib.so.3.
 I tried to install the libdxflib library but nothing happened.

**Comment by ilya301444** [comment]
Everything is installed and running

**Comment by eyal0** [comment]
It works?

**Comment by mikhailmihalkov** [comment]
> It works?

Yes, it works on ubuntu only to install as a program cannot be
[gerbv.zip](https://github.com/gerbv/gerbv/files/9890861/gerbv.zip)
 forgotten from one file.... How to make a deb package is not known....

**Comment by eyal0** [comment]
I also do not know how.

**Comment by mikhailmihalkov** [comment]
https://launchpad.net/ubuntu/+source/gerbv/2.9.2-1

**Comment by eyal0** [comment]
It appears that it was made by Carsten Schoenert.  I do not know him.  Maybe he can help us?

I do not make the package, I don't know how.  :frowning_face:

**Comment by mikhailmihalkov** [comment]
I wrote off the programmer promised to help) collect rpm package

**Comment by ooxi** [comment]
The fix by @eyal0 will be included in the next release

**Comment by mikhailmihalkov** [comment]
eyal0  With each new launch of the application on win 10, the settings of the unit of measure are not saved, the selection of the render is fast normal, etc... Is this a bug or is it intended?

**Comment by ooxi** [comment]
@mikhailmihalkov fix by @eyal0  for this issue has been included in release [v2.9.4](https://github.com/gerbv/gerbv/releases/tag/v2.9.4), thanks for the report!


> eyal0 With each new launch of the application on win 10, the settings of the unit of measure are not saved, the selection of the render is fast normal, etc... Is this a bug or is it intended?

It's best to report this as a separate issue
